python/ray/data/_internal/logical/interfaces.py

```diff
@@ -0,0 +1,123 @@
+
+    def set_input_dependencies(self, input_dependencies: List["LogicalOperator"]):
+        self._input_dependencies = input_dependencies
+
+    def set_legacy_stage(self, stage: Stage):
+        self._stage = stage
+
+
+class LogicalPlan:
+    """The plan with a DAG of logical operators."""
+
+    def __init__(self, dag: LogicalOperator):
+        self._dag = dag
+
+    def add(self, operator: LogicalOperator) -> "LogicalPlan":
+        """Add the operator into DAG of logical operators."""
+        assert self._dag is not None
+        operator.set_input_dependencies([self._dag])
+        self._dag = operator
+        return self
+
+    def get_execution_dag(self) -> PhysicalOperator:
+        """Get the DAG of physical operators to execute.
+
+        This process has 3 steps:
+        (1).logical optimization: optimize logical plan.
+        (2).planning: convert logical to physical plan.
+        (3).physical optimization: optimize physical plan.
+        """
+        logical_optimizer = LogicalOptimizer()
+        optimized_plan = logical_optimizer.optimize(self)
+        physical_plan = logical_optimizer.plan(optimized_plan)
+        final_plan = PhysicalOptimizer().optimize(physical_plan)
+        return final_plan._dag
+
+
+class PhysicalPlan:
+    """The plan with a DAG of physical operators."""
+
+    def __init__(self, dag: PhysicalOperator):
+        self._dag = dag
+
+
+class LogicalOptimizer:
+    """The optimizer for logical plan."""
+
+    def __init__(self):
+        self._rules = []
+
+    def optimize(self, plan: LogicalPlan) -> LogicalPlan:
+        """Optimize logical plan with a list of rules."""
+        for rule in self._rules:
+            plan = rule.apply(plan)
+        return plan
+
+    def plan(self, logical_plan: LogicalPlan) -> PhysicalPlan:
+        """Convert logical plan into physical plan."""
+        return PhysicalPlan(self._plan_op(logical_plan._dag))
+
+    def _plan_op(self, op: LogicalOperator) -> PhysicalOperator:
+        """Convert logical operators to physical operators in post-order."""
+        # Plan the input dependencies first.
+        physical_children = []
+        for child in op.input_dependencies:
+            physical_children.append(self._plan_op(child))
+
+        if op.name == "Read":
+            assert not physical_children
+            physical_op = _blocks_to_input_buffer(op._blocks, owns_blocks=True)
+        elif op.name == "MapBatches":
+            assert len(physical_children) == 1
+            physical_op = _stage_to_operator(op._stage, physical_children[0])
+        else:
+            raise ValueError(f"Found unknown logical operator during planning: {op}")
+        return physical_op
+
+
+class PhysicalOptimizer:
+    """The optimizer for physical plan."""
+
+    def __init__(self):
+        self._rules = []
+
+    def optimize(self, plan: PhysicalPlan) -> PhysicalPlan:
+        """Optimize physical plan with a list of rules."""
+        for rule in self._rules:
```

Comment:
What rules would we put here initially?

Reply:
Stage fusion should be put here initially for physical optimization. This is similar to Spark's whole stage code-gen rule (collapse multiple operators into one whole stage code-gen operator), is part of physical optimization ([Spark code](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L440)).